### PR TITLE
Add autoupdate feature

### PR DESCRIPTION
**Pull request by Chronial**
That was a bit of work, but it's done :).

There are some ways how this could be improved on the server-side:
* The executable just grew to 1MB because the library that extracts rar files is 700kb. If you would pack your releases as zip files, we could use core library code instead.
* Building the download url from the version string does not feel very good. We should send the urls in the version file. Probably line 3 and 4?

It also very important that you do not add the version to the filename of the executable inside the `rar` in future releases, as that filename will be wrong after the first autoupdate.

If you want to make the relevant changes on the server side, do them and tell me, then I will update this code.

### CODE REVIEW

**Comment by Wulf2k** [comment]
What are your thoughts on dropping the compression entirely and just have them download an EXE?

**Comment by Chronial** [comment]
I think that would cause problems with the filename? With the auto-update feature, all downloaded EXEs should have the same name as to not cause confusion after the update.

File: DaS-PC-MPChan/UpdateWindow.vb [hunk 1]
```diff
@@ -94,12 +94,13 @@ Public Class UpdateWindow
...
-                If ignoreExtensions.Contains(Path.GetExtension(reader.Entry.Key)) Then
+                Dim entryExtension = Path.GetExtension(reader.Entry.Key).ToLower()
+                If ignoreExtensions.Contains(entryExtension) Then
                     Continue While
                 End If
                 If (extractedContents IsNot Nothing Or
                         reader.Entry.IsDirectory Or
-                        Not Path.GetExtension(reader.Entry.Key) = ".exe") Then
+                        Not entryExtension = ".exe") Then
                     Throw New ApplicationException("Unkown structure, please update manually")
```
Comment (Wulf2k) [inline DaS-PC-MPChan/UpdateWindow.vb]: "Unk-n-own"

**Comment by Wulf2k** [comment]
The downloaded EXE doesn't have to have the same name as it does on the server.  Just download it as "DSCM.exe".

I'd prefer that over adding a bunch of extra libraries to the EXE.  I dunno why, but 1 MB just seems a bit obscene for what it does.

**Comment by Chronial** [comment]
I'm afraid you misunderstood both of my points :)

1. If you pack the file as zip, we don't need a library – .Net can extract zip files.
2. The problem is that we replace the file of the running instance. So if the file was called `DSCM-2016-05-12.exe` before the update, it will be called that after the update, too – which would be misleading. So the point is that the file should not have a version in its name *before* the update.

I don't think it's a good idea to download the new version with a different name than the running file, as it will break shortcuts.

**Comment by Wulf2k** [comment]
Ah, I follow on both points now.

Ok, I'm fine with switching to ZIPs and maintaining a constant EXE name.  The days of needing to immediately know which version somebody was running are past, and I can ask in any case that it matters.

Line 3 of the version file can be a stable download link, and line 4 can be a testing download link.

**Comment by Chronial** [comment]
Great :). It would be helpful if you could put the files on the server like this so I can then test the code new code.

Regarding getting the version: I don't know if you noticed, but I made a small change to the build setup, so every build has a different file version number. You can see an example of this in the report in #69. The user was using DSCM build `1.0.5976.40573`. That number is going to be unique for every build.

**Comment by Wulf2k** [comment]
Yup.

Though a table showing which build had which features might make that more
useful :)

I'll see if i can get the server stuff sorted out 6 hours from now.

**Comment by Chronial** [comment]
We could improve the versioning situation in multiple ways:

1. You can tag releases in git. This way you can easily see in github which release had which features/commits.
2. If you switch to a versioning scheme that is compatible with the .Net assembly versioning (`x.x.x.x`, ideally `x.x.x`, so we still have the last field for build numbers), you could have the actual version in that field.

File: DaS-PC-MPChan/MainWindow.vb [hunk 2]
```diff
@@ -322,7 +322,7 @@ Public Class MainWindow
             Dim uri = "http://wulf2k.ca/pc/das/dscm-ver.txt"
             Dim content As String = Await client.DownloadStringTaskAsync(uri)
 
-            Dim lines() As String = content.Split({Chr(10), Chr(13)})
+            Dim lines() As String = content.Split({vbCrLf, vbLf}, StringSplitOptions.None)
```
Comment (Wulf2k) [inline DaS-PC-MPChan/MainWindow.vb]: Was there anything wrong with the file.readalllines that handled this before?

Comment (Chronial) [inline DaS-PC-MPChan/MainWindow.vb]: Yes, it only works on files, but it's unnecessary to store anything in a file here.

**Pull request merged**